claws-mail{,-gtk3}: 3.17.7->3.17.8 and refactoring

oxzi opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
claws-mail{,-gtk3}: major refactoring

Previously, the configurable arguments were neither complete nor named
according to the configure.ac file. Likewise, the values did not
correspond to the defaults, but rather to a personal preference.

This has now been changed to enable the arguments which are enabled in
the configure.ac file. Also the variable names have been adjusted. For
compatibility the old parameters also exist.

Next to the claws-mail package is the "experimental" claws-mail-gtk3
package for the non official gtk3 git branch. This package started as an
almost one-to-one copy of the claws-mail derivation which small
modifications. This package was of course not updated.

This has also been changed so that both packages are built from the same
derivative.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

oxzi committed:
claws-mail{,-gtk3}: 3.17.7->3.17.8 and refactoring

Update Claws Mail to its latest version and perform a major refactoring.

Previously, the configurable arguments were neither complete nor named
according to the configure.ac file. Likewise, the values did not
correspond to the defaults, but rather to a personal preference.

This has now been changed to enable the arguments which are enabled in
the configure.ac file. Also the variable names have been adjusted. For
compatibility the old parameters also exist.

Next to the claws-mail package is the "experimental" claws-mail-gtk3
package for the non official gtk3 git branch. This package started as an
almost one-to-one copy of the claws-mail derivation which small
modifications. This package was of course not updated.

This has also been changed so that both packages are built from the same
derivative.

doronbehar commented on pkgs/applications/networking/mailreaders/claws-mail/default.nix:
```diff
@@ -1,97 +1,171 @@
   '';
 
-  nativeBuildInputs = [ autoreconfHook pkgconfig wrapGAppsHook python.pkgs.wrapPython ];
+  nativeBuildInputs = [ autoreconfHook pkgconfig bison flex wrapGAppsHook python.pkgs.wrapPython ];
```
How come all of a sudden?

oxzi replied:
The tarballs contain some generated content, like precompiled translations. Those needed to be generated when compiled from `git`. The old `-gtk3` package had those inputs included and they did the job. Thus, I added them to the _merged_ package description.

doronbehar replied:
You can add it only if gtk3 is used.

oxzi replied:
Now we are also sourcing GTK2 from git because reasons.

doronbehar commented on pkgs/applications/networking/mailreaders/claws-mail/default.nix:
```diff
@@ -1,97 +1,171 @@
+    ++ optional  enablePluginRssyl libxml2
+    ++ optionals enablePluginPgp [ gnupg gpgme ]
+    ++ optional  enablePluginTnefParse libytnef
+    ++ optional  enablePluginVcalendar libical;
 
```
Nitpick, that might make future diffs cleaner by a bit:

```suggestion
    ++ optional  enablePluginVcalendar libical
  ;
```

doronbehar commented on pkgs/applications/networking/mailreaders/claws-mail/default.nix:
```diff
@@ -1,97 +1,171 @@
+    ++ optional  (!enablePluginSpamassassin) "--disable-spamassassin-plugin"
+    ++ optional  (!enablePluginSpamReport) "--disable-spam_report-plugin"
+    ++ optional  (!enablePluginTnefParse) "--disable-tnef_parse-plugin"
+    ++ optional  (!enablePluginVcalendar) "--disable-vcalendar-plugin";
 
```
```suggestion
    ++ optional  (!enablePluginVcalendar) "--disable-vcalendar-plugin"
  ;
```

doronbehar reviewed (commented):
The pros use [`lib.strings.enableFeature`](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.strings.enableFeature), but it's not that suprt important :). Unless, you would want to go all the way, and do something similar to what mpd does:

https://github.com/NixOS/nixpkgs/blob/62e614262ad4b9df3d40693350f516794e95ecda/pkgs/servers/mpd/default.nix#L35-L89

And use the word after `--disable-` as the attribute name.

All of this requires more Nix experience which is not mandatory, but is encouraged :). See another example:

https://github.com/doronbehar/nixpkgs/blob/c4ffa921a7ff5ef25f2ab867562fc29d76c622ef/pkgs/applications/radio/gnuradio/default.nix#L50

oxzi commented:
@doronbehar: Thanks a lot for your feedback and please excuse the delay on my side.

I updated the PR to use `lib.strings.enableFeature`. This resulted in a _forced_ enabling which brought up some errors which were silenced before. For example the GTK3 version requires Python 3 and other Python packages. I adjusted the PR accordingly.

Furthermore, a new Claws release came along the way, which I included in this PR.

doronbehar commented on pkgs/applications/networking/mailreaders/claws-mail/default.nix:
```diff
@@ -1,104 +1,187 @@
+  gtk2src = {
+    rev = version;
+    sha256 = "0l4f8q11iyj8pi120lrapgq51k5j64xf0jlczkzbm99rym752ch5";
+  };
```
On a 2nd thought. gtk2 is deprecated, why should we even bother supporting it?

oxzi replied:
Because this is the official release of Claws Mail. The GTK3 branch is very experimental and its not tested whatsoever.

doronbehar commented on pkgs/applications/networking/mailreaders/claws-mail/default.nix:
```diff
@@ -1,104 +1,187 @@
+      [ enableSvg "svg" ]
 
-  enableParallelBuilding = true;
+      [ enablePluginAcpiNotifier "acpi_notifier-plugin" ]
+      [ enablePluginAddressKeeper "address_keeper-plugin" ]
+      [ enablePluginArchive "archive-plugin" ]
+      [ enablePluginAttRemover "att_remover-plugin" ]
+      [ enablePluginAttachWarner "attachwarner-plugin" ]
+      [ enablePluginBogofilter "bogofilter-plugin" ]
+      [ enablePluginBsfilter "bsfilter-plugin" ]
+      [ enablePluginClamd "clamd-plugin" ]
+      [ enablePluginDillo "dillo-plugin" ]
+      [ enablePluginFetchInfo "fetchinfo-plugin" ]
+      [ enablePluginLibravatar "libravatar-plugin" ]
+      [ enablePluginLitehtmlViewer "litehtml_viewer-plugin" ]
+      [ enablePluginMailmbox "mailmbox-plugin" ]
+      [ enablePluginManageSieve "managesieve-plugin" ]
+      [ enablePluginNewMail "newmail-plugin" ]
+      [ enablePluginNotification "notification-plugin" ]
+      [ enablePluginPdfViewer "pdf_viewer-plugin" ]
+      [ enablePluginPerl "perl-plugin" ]
+      [ enablePluginPython "python-plugin" ]
+      [ enablePluginPgp "pgpcore-plugin" ]
+      [ enablePluginPgp "pgpmime-plugin" ]
+      [ enablePluginPgp "pgpinline-plugin" ]
+      [ enablePluginRssyl "rssyl-plugin" ]
+      [ enablePluginSmime "smime-plugin" ]
+      [ enablePluginSpamassassin "spamassassin-plugin" ]
+      [ enablePluginSpamReport "spam_report-plugin" ]
+      [ enablePluginTnefParse "tnef_parse-plugin" ]
+      [ enablePluginVcalendar "vcalendar-plugin" ]
```
This should get a lot cleaner and nicer with some Nix attributes and functions applied. Start with this up above:

```nix
pluginsDeps = {
  acpi_notifier = [];
  archive = [ /*..*/ ];
  att_remover = [];
  # ...
};
```

Then use [`mapAttributesToList`](https://nixos.org/manual/nixpkgs/stable/#function-library-lib.attrsets.mapAttrsToList) to map the plugins enabled, to list of `configureFlags` (with `enableFeature`) and a list of `pluginsDeps`.

oxzi replied:
You've got a point there. I will take a look at this later / tomorrow. Thanks!

SuperSandro2000 replied:
@oxzi Are you still planning to do this?

doronbehar replied:
Perhaps it'd be worth re-reviewing this, for the sake of the update, and revisit in the future the idea at using such Nix attributes..

oxzi commented:
Please excuse this stalled pull request. Both the successive change requests and the fact that I no longer use Claws as my go-to email client have caused me to stop actively working on this pull request. Thus, I am going to close this one. Maybe someone else will keep up this work. Thanks to all the reviewers.
